internal/controller/clientprofilemapping_controller.go

```diff
@@ -18,45 +18,184 @@ package controller
-	_ = log.FromContext(ctx)
+// A local reconcile object tied to a single reconcile iteration
+type ClientProfileMappingReconcile struct {
+	ClientProfileMappingReconciler
 
-	// TODO(user): your logic here
+	ctx                      context.Context
+	log                      logr.Logger
+	req                      ctrl.Request
+	clientProfileMappingList csiv1a1.ClientProfileMappingList
+}
 
-	return ctrl.Result{}, nil
+// csiClusterMappingRecord represent the structure to serialize a csi mapping
+// record in Ceph CSI's config, configmap
+type csiClusterMappingRecord struct {
+	ClusterIdMapping map[string]string   `json:"clusterIdMapping,omitempty"`
+	RbdPoolIdMapping []map[string]string `json:"RBDPoolIDMapping,omitempty"`
 }
 
 // SetupWithManager sets up the controller with the Manager.
 func (r *ClientProfileMappingReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	genChangedPredicate := predicate.GenerationChangedPredicate{}
+
 	return ctrl.NewControllerManagedBy(mgr).
-		For(&csiv1alpha1.ClientProfileMapping{}).
+		For(&csiv1a1.ClientProfileMapping{}).
+		// TODO: This watch is probalmatic, it will trigger a reconcile for each
+		// mapping resource that exists in the configmap namespace whenever the configmap
+		// is changed. This is unnececary as this reconile build its desired state based
+		// on all mapping resources within a single reconcile
+		Owns(
+			&corev1.ConfigMap{},
+			builder.MatchEveryOwner,
+			builder.WithPredicates(
+				utils.NamePredicate(utils.CsiConfigVolume.Name),
+				genChangedPredicate,
+			),
+		).
 		Complete(r)
 }
+
+func (r *ClientProfileMappingReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := ctrllog.FromContext(ctx)
+	log.Info("Starting reconcile iteration for ClientProfileMapping", "req", req)
+
+	reconcileHandler := ClientProfileMappingReconcile{}
+	reconcileHandler.ClientProfileMappingReconciler = *r
+	reconcileHandler.ctx = ctx
+	reconcileHandler.log = log
+	reconcileHandler.req = req
+
+	err := reconcileHandler.reconcile()
+	if err != nil {
+		log.Error(err, "ClientProfileMapping reconciliation failed")
+	} else {
+		log.Info("ClientProfileMapping reconciliation completed successfully")
+	}
+	return ctrl.Result{}, err
+}
+
+func (r *ClientProfileMappingReconcile) reconcile() error {
+	// This controller behave a differently then other controller. Because of the lack of uniqueness of the mapping
+	// information between 2 or more mapping resources, there is no way to update the csi mapping information without
+	// building it from the gournd up on every reconcile. To do that we ignore the specific of the subject resource
+	// and instead load all mapping resources within the namespace.
+	if err := r.List(r.ctx, &r.clientProfileMappingList, client.InNamespace(r.req.Namespace)); err != nil {
+		r.log.Error(err, "Failed listing ClientProfileMapping CRs in namespace", "namespace", r.req.Namespace)
+		return err
```

Comment:
can we record the CRName in the csiClusterMappingRecord configmap to avoid list operation?

Reply:
How does that help? The list operation is not here to find the record but to consilidate the fact that many CRs might be adding to the same record, even the same data. I am not aware of any way to calculate the final state of a record without taking into account data from all relevant CRs

Reply:
got it now, lets keep it as it is.

Reply:
Fixed